File: framework/languages/java/samples/java/DeliveryDispatch/Server/CourierSpotNode/src/main/java/systems/zlink/samples/deliverydispatch/server/courierspotnode/handlers/OfferDeliveryRouteHandler.java
package systems.zlink.samples.deliverydispatch.server.courierspotnode.handlers;

import static systems.zlink.framework.ZLinkAwait.await;

import systems.zlink.framework.actors.ZLinkActorManager;
import systems.zlink.framework.actors.ZLinkActorRef;
import systems.zlink.framework.channels.ZLinkRouteRequestContext;
import systems.zlink.framework.channels.ZLinkRouteRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleNames;
import systems.zlink.samples.deliverydispatch.server.courierspotnode.ActorDirectory;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;

@ZLinkHandlerGroup("courier-actor-node")
public final class OfferDeliveryRouteHandler
    implements ZLinkRouteRequestHandler<Messages.OfferDelivery, Messages.OfferDeliveryResult> {
    private final ZLinkActorManager actors;
    private final ActorDirectory directory;

    public OfferDeliveryRouteHandler(
        ZLinkActorManager actors,
        ActorDirectory directory) {
        this.actors = actors;
        this.directory = directory;
    }

    @Override
    public Messages.OfferDeliveryResult handle(
        Messages.OfferDelivery request,
        ZLinkRouteRequestContext context) {
        ZLinkActorRef actor = await(actors.getOrCreate(
            request.courierId(),
            SampleNames.CourierActorType,
            request));
        return directory.require(actor.actorId()).offer(request);
    }
}

File: framework/languages/java/samples/kotlin/DeliveryDispatch/Server/CourierSpotNode/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/courierspotnode/handlers/OfferDeliveryRouteHandler.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.courierspotnode.handlers

import systems.zlink.framework.kotlin.await
import systems.zlink.framework.kotlin.ZLinkSuspendingSpotRequestHandler
import systems.zlink.framework.actors.ZLinkActorManager
import systems.zlink.samples.kotlin.deliverydispatch.server.courierspotnode.ActorDirectory
import systems.zlink.samples.kotlin.deliverydispatch.server.courierspotnode.spots.CourierEntrySpot
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.OfferDeliveryReq
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.OfferDeliveryRes

class OfferDeliveryRouteHandler(
    private val actors: ZLinkActorManager,
    private val directory: ActorDirectory,
) : ZLinkSuspendingSpotRequestHandler<CourierEntrySpot, OfferDeliveryReq, OfferDeliveryRes> {
    override suspend fun handle(
        spot: CourierEntrySpot,
        request: OfferDeliveryReq,
    ): OfferDeliveryRes {
        val actor = actors.find(request.courierId).await()
            .orElseThrow { IllegalStateException("Courier actor is not bound: ${request.courierId}") }
        return directory.require(actor.actorId()).offer(request)
    }
}
